src/python/pants/init/extension_loader.py

```diff
@@ -123,25 +129,60 @@ def load_build_configuration_from_source(
     """
     # NB: Backends added here must be explicit dependencies of this module.
     backend_packages = FrozenOrderedSet(["pants.core", "pants.backend.project_info", *backends])
+    templated_backends = templated_backends or {}
+
     for backend_package in backend_packages:
-        load_backend(build_configuration, backend_package)
+        load_backend(
+            build_configuration,
+            backend_package,
+            templating_config=templated_backends.get(backend_package),
+        )
 
 
-def load_backend(build_configuration: BuildConfiguration.Builder, backend_package: str) -> None:
+def _load_register_module(backend_package: str):
+    module_source = backend_package + ".register"
+    try:
+        module = importlib.import_module(module_source)
+    except ImportError as ex:
+        traceback.print_exc()
+        raise BackendConfigurationError(f"Failed to load the {module_source} backend: {ex!r}")
+    return module, module_source
+
+
+def _generate_backend_module_from_template(backend_package: str, cfg: TemplatedBackendConfig):
+    try:
+        backend_generator_module = importlib.import_module(cfg.template)
+    except ImportError as ex:
+        traceback.print_exc()
+        raise BackendConfigurationError(
+            f"Failed to load the {cfg.template} backend template module: {ex!r}"
+        )
+    module = backend_generator_module.generate(backend_package, cfg.kwargs)
+    module_source = f"{module} generated from {cfg.template}"
+    return module, module_source
```

Comment:
These two functions return objects with the same an interface: `module` is something with zero-arg callable attributes like `target_types`, `rules` and `module_source` is a string describing the module for use in error messages.